File: public/packages/dotnet/geo/abstractions/NameResolution/IGeoNameResolver.cs
// -----------------------------------------------------------------------
// <copyright file="IGeoNameResolver.cs" company="DCSV">
// Copyright (c) DCSV. Licensed under the Apache License, Version 2.0.
// </copyright>
// -----------------------------------------------------------------------

namespace DcsvIo.D2.Geo.Abstractions.NameResolution;

using DcsvIo.D2.Result;

/// <summary>
/// Contract for resolving a free-form place-name string (from a 3rd-party
/// source — WhoIs response, IP-geolocation enrichment, vendor API, user
/// input) into the corresponding catalog record. Used at the integration
/// boundary; never called from domain-handler code (typed-handler-input
/// discipline).
/// </summary>
/// <remarks>
/// <para>
/// <b>Fail-closed semantics.</b> Implementations MUST be fail-closed:
/// when input is null / empty / whitespace / ambiguous / too short for a
/// fuzzy pass, OR when multiple entities match at the same score within
/// a pass, the resolver returns <see cref="D2Result.NotFound"/> rather
/// than guessing. The downstream consumer (in the WhoIs enrichment flow)
/// preserves the raw upstream string in the audit trail, so an
/// unresolved outcome is safe — a wrong fuzzy match is not.
/// </para>
/// <para>
/// <b>Cascade.</b> The resolution algorithm is a four-pass cascade with
/// per-pass minimum-input-length guards:
/// </para>
/// <list type="number">
///   <item>
///     <description>Pass 1 (exact) — no minimum input length. Always runs.</description>
///   </item>
///   <item>
///     <description>Pass 2 (startsWith) — skip if <c>q.length &lt; 4</c>.
///     Short prefixes are too promiscuous.</description>
///   </item>
///   <item>
///     <description>Pass 3 (contains) — skip if <c>q.length &lt; 5</c>.
///     Short substrings overmatch.</description>
///   </item>
///   <item>
///     <description>Pass 4 (Levenshtein) — skip if <c>q.length &lt; 5</c>;
///     within-pass <c>maxDistance &lt;= min(2, floor(q.length / 5))</c>.
///     Fuzzy on short inputs collapses distinct entities.</description>
///   </item>
/// </list>
/// <para>
/// The first matching pass wins; later passes do not run. Ambiguity within
/// a pass (multiple matches at the same score) returns
/// <see cref="D2Result.NotFound"/>. Catalog uniqueness is enforced at
/// codegen time via <c>D2GEO010</c>, so the determinism risk of
/// "first match wins" is eliminated at the source.
/// </para>
/// </remarks>
public interface IGeoNameResolver
{
    /// <summary>
    /// Resolves a free-form country-name string to a typed
    /// <see cref="Country"/> identifier via the fail-closed cascade.
    /// </summary>
    /// <param name="name">
    /// The free-form country name. May be null / empty / whitespace —
    /// implementations MUST validate input before performing any lookup and
    /// return <c>D2Result.ValidationFailed</c> for null / empty / whitespace,
    /// or <c>D2Result.NotFound</c> at implementer's discretion; either way
    /// the caller never sees a wrong-Country answer for bad input.
    /// </param>
    /// <returns>
    /// <c>D2Result.Ok(country)</c> on unambiguous match; <c>NotFound</c> on
    /// ambiguity, too-short input for any-pass match, or no-match at all
    /// passes; <c>ValidationFailed</c> on null / empty / whitespace input.
    /// </returns>
    D2Result<Country> TryResolveCountryByName(string name);

    /// <summary>
    /// Resolves a free-form subdivision-name string to the catalog's
    /// <see cref="Subdivision"/> record via the fail-closed cascade,
    /// scoped to <paramref name="parentCountry"/>. The parent-country
    /// scope is mandatory — "Georgia" is BOTH a country (GE) AND a US
    /// state (US-GA), and the resolver MUST NOT silently pick one.
    /// </summary>
    /// <param name="name">The free-form subdivision name.</param>
    /// <param name="parentCountry">The country whose subdivisions to search.</param>
    /// <returns>
    /// <c>D2Result.Ok(subdivision)</c> on unambiguous match;
    /// <c>NotFound</c> on ambiguity, too-short input, no match in that
    /// country's catalog, or coverage gap; <c>ValidationFailed</c> on
    /// null / empty / whitespace input.
    /// </returns>
    D2Result<Subdivision> TryResolveSubdivisionByName(
        string name, Country parentCountry);
}

File: public/packages/typescript/geo/abstractions/src/name-resolution/i-geo-name-resolver.ts
// -----------------------------------------------------------------------
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) DCSV
// -----------------------------------------------------------------------

import type { D2Result } from "@dcsv-io/d2-result";

import type { Country } from "../generated/country.g.js";
import type { Subdivision } from "../generated/subdivision.g.js";

/**
 * Mirror of .NET `DcsvIo.D2.Geo.Abstractions.NameResolution.IGeoNameResolver`
 * — the cascade-based free-form text → entity resolver for countries and
 * subdivisions. Resolves `"United States"` → the `Country` record,
 * `"California"` (within US) → the `Subdivision` record, etc.
 *
 * Cascade order (Pass 1 → Pass 4):
 *
 * 1. **Pass 1** — exact normalized match (always runs, no min input length).
 * 2. **Pass 2** — `startsWith` prefix match (skipped if `q.length < 4`).
 * 3. **Pass 3** — `contains` substring match (skipped if `q.length < 5`).
 * 4. **Pass 4** — Levenshtein fuzzy match (skipped if `q.length < 5`;
 *    per-pass `maxDistance ≤ min(2, floor(q.length / 5))`).
 *
 * Each pass runs over the catalog's matchable name fields (`displayName`,
 * `officialName`, endonym fields, alpha-3 / numeric codes for countries) and
 * short-circuits the cascade as soon as a pass produces ≥1 match. The
 * `normalize()` pipeline at `./name-normalizer.ts` produces the comparison
 * key on both sides.
 *
 * **Fail-closed semantics**: implementations MUST
 * return `D2Result.notFound(...)` rather than guessing when input is
 * empty / whitespace / too short for the cascade pass that would have
 * matched it / ambiguous (multiple matches at the same score within a
 * pass) / no-match at all passes. Rationale: incorrect resolution
 * silently corrupts downstream data; unresolved outcomes are safe
 * because the raw upstream string is preserved in the audit trail.
 *
 * **Cache-aside semantics**: implementations build their normalized-name
 * → record map lazily on first call. First lookup is O(n) over the
 * catalog; subsequent lookups are O(1) against the cached map. Mirrors
 * .NET `DefaultGeoNameResolver` semantics.
 *
 * `D2Result<T>` (from `@dcsv-io/d2-result`) is the cross-language result
 * envelope — mirrors .NET `D2Result<T>` byte-for-byte over the wire.
 *
 * Both .NET and TS resolvers return full records (never bare codes) so
 * callers get the natural fields immediately without a second lookup
 * — codes are available as a property on each returned record.
 */
export interface IGeoNameResolver {
  /**
   * Resolve a free-form country name to the catalog `Country` record.
   *
   * @example
   * `tryResolveCountryByName("United States")` →
   *   `D2Result.ok(Countries.US)`
   * `tryResolveCountryByName("IR")` → `D2Result.notFound(...)` (too short
   *   for Pass 2-4 cascade, no Pass-1 exact hit).
   */
  tryResolveCountryByName(name: string): D2Result<Country>;

  /**
   * Resolve a free-form subdivision name within the scope of
   * `parentCountry` to the catalog `Subdivision` record. Scoping is
   * REQUIRED — `"Western"` is ambiguous globally but unique within most
   * countries; `"Georgia"` is BOTH a country (GE) AND a US state (US-GA),
   * and the resolver MUST NOT silently pick one.
   *
   * @example
   * `tryResolveSubdivisionByName("California", Countries.US)`
   *   → `D2Result.ok(Subdivision{ iso31662Code: "US-CA", ... })`
   * `tryResolveSubdivisionByName("Carolina", Countries.US)`
   *   → `D2Result.notFound(...)` (Pass-3 contains-ambiguity — matches
   *      both North + South Carolina; fail-closed).
   */
  tryResolveSubdivisionByName(
    name: string,
    parentCountry: Country,
  ): D2Result<Subdivision>;
}
